README.txt
= BasicServer

* http://github.com/illuminerdi/basic_server

== DESCRIPTION:

A simple HTTP server that works under GServer and TCPServer

== FEATURES/PROBLEMS:

2 HTTP Servers (really basic HTTP 0.9/GET only):

    * gserver for thread spawning per connection
    * tcpserver w/ prespawned thread pool

requirements:

    * craft your own request / response objects
    * servelets of your own design. start with time of the day app.
    * static files from the filesystem.
    * html files served plain
    * html.erb files should render via erb.
    * no caching, no clever.

Bare minimum HTTP protocol:

    * request:

GET /index.html HTTP/1.0 CRLF CRLF

    * response:

HTTP/1.1 200 OK CRLF Date: Thu, 07 May 2009 01:06:01 GMT CRLF Last-Modified: Mon, 19 Jan 2009 23:02:19 GMT CRLF Content-Length: 3750 CRLF Content-Type: text/html CRLF CRLF …content…

    * other responses are:
    * 304 redirected
    * 404 missing
    * 500 app error


== SYNOPSIS:

  # you can start up either server directly from the command line

  # For the GServer implementation:
  $ bin/basic_server gserver

  # For the TCPServer implementation:
  $ bin/basic_server tcpserver

== REQUIREMENTS:

* ruby 1.8.6

== INSTALL:

* Download, use.

== LICENSE:

(The MIT License)

Copyright (c) 2009 Joshua Clingenpeel

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 1.0.0 / 2009-05-10

* 1 major enhancement

  * Birthday!



test/test_basic_server.rb
require 'test/unit'
require 'basic_server'

class TestBasicServer < Test::Unit::TestCase
  def setup
    @request = BasicServer::Request.new("GET / HTTP/1.1\r\n\r\n")
    @not_found = BasicServer::Request.new("GET /notfound.htm HTTP/1.1\r\n\r\n")
    @erb = BasicServer::Request.new("GET /test.html.erb HTTP/1.1\r\n\r\n")
    @time = BasicServer::Request.new("GET /time HTTP/1.1\r\n\r\n")
    @error = BasicServer::Request.new("GET /error HTTP/1.1\r\n\r\n")
    @quotes = BasicServer::Request.new("GET /random_quote HTTP/1.1\r\n\r\n")
  end

  def test_request_has_a_method_and_path
    assert_equal "public_html/index.htm", @request.path.to_s
    assert_equal "GET", @request.method
  end

  def test_ok_header
    header = @request.create_header(200)
    assert_match(/HTTP\/0.9 200 OK/, header)
    assert_match(/text\/html/, header)
  end

  def test_ok_servlet_header_with_content_type
    header = @time.create_header(200)
    assert_match(/text\/plain/, header)
  end

  def test_not_found_header
    header = @request.create_header(404)
    assert_match(/HTTP\/0.9 404 Not Found/, header)
    assert_match(/text\/html/, header)
  end

  def test_exception_header
    header = @request.create_header(500)
    assert_match(/HTTP\/0.9 500 Internal Server Error/, header)
    assert_match(/text\/html/, header)
  end

  def test_time_servlet
    assert_match(/Current server time/, @request.time)
  end

  def test_error_servlet
    assert_raise Exception do
      @request.error
    end
  end

  def test_quotes_servlet
    assert_match(/-Oscar Wilde/, @request.random_quote)
  end

  def test_html_response_http_code
    response = @request.respond
    assert_equal "HTTP/0.9 200 OK", get_status(response)
  end

  def test_html_response_header_info
    response = @request.respond
    response.pop
    response[0].split("\r\n").each {|line|
      assert_match(/HTTP|Server|Date|Last-Modified|Content-Length|Content-Type/, line)
    }
  end

  def test_erb_response
    response = @erb.respond
    assert_equal "HTTP/0.9 200 OK", get_status(response)
    assert_match(/<html>/, response.last)
  end

  def test_time_servlet_response
    response = @time.respond
    assert_equal "HTTP/0.9 200 OK", get_status(response)
    assert_match(/Current server time is:/, response.last)
  end

  def test_quotes_servlet_response
    response = @quotes.respond
    assert_equal "HTTP/0.9 200 OK", get_status(response)
    assert_match(/-Oscar Wilde/, response.last)
  end

  def test_error_servlet_response
    # we're passing the exception back through to the calling module method, which will handle the http 500
    assert_raise(Exception) do
      response = @error.respond
    end
  end

  def test_file_not_found
    response = @not_found.respond
    assert_equal "HTTP/0.9 404 Not Found", get_status(response)
  end

  def get_status(response)
    response[0].split("\r\n").first
  end
end

lib/basic_server.rb
#!/usr/bin/env ruby -w

require 'gserver'
require 'socket'
require 'pathname'
require 'erb'
require 'thread'

module BasicServer
  VERSION = '1.0.0'
  RESPONSES = {
    200 => "OK",
    404 => "Not Found",
    500 => "Internal Server Error"
  }
  HTTP_VERSION = "HTTP/0.9"
  FILE_404 = "public_html/404.html"
  FILE_500 = "public_html/500.html"

  class Request
    attr_reader :method, :path

    def initialize(request)
      @method, path = request.match(/^(\w+)\s(\S+)\s(\S+)/)[1,2]
      path = "/index.htm" if path.match(/^\/$/)
      @path = Pathname.new("public_html#{path}")
    end

    def create_header rnum
      type = case @path.extname
        when /^(\.htm|\.html|\.erb)/ then 'text/html'
        else 'text/plain'
      end
      header = "#{HTTP_VERSION} #{rnum} #{RESPONSES[rnum]}\r\n"
      header += "Date: #{Time.now}\r\n"
      header += "Server: BasicServer\r\n"
      if @path.exist? && @path.file?
        header += "Last-Modified: #{@path.mtime}\r\n"
        header += "Content-Length: #{@path.size}\r\n"
      end
      header += "Content-Type: #{type}\r\n\r\n"
    end

    def respond
      response = []
      template = @path.split.last.to_s
      if self.respond_to?(template.to_sym)
        # we are a servlet
        response << create_header(200)
        response << self.send(template)
        return response
      end

      unless @path.exist? && @path.file?
        response << create_header(404)
        response << Pathname.new(FILE_404).read
        return response
      end

      response << create_header(200)
      case @path.extname
      when '.erb'
        response << ERB.new(@path.read.gsub(/^  /, '')).result
      when /^(\.htm|\.html)/
        response << @path.read
      end
    end

    def time
      "Current server time is: #{Time.now.to_s}"
    end

    def error
      raise Exception, "You should be seeing an error."
    end

    def random_quote
      quotes = [
        "A little sincerity is a dangerous thing, and a great deal of it is absolutely fatal. -Oscar Wilde",
        "A man can be happy with any woman as long as he does not love her. -Oscar Wilde",
        "Always forgive your enemies; nothing annoys them so much. -Oscar Wilde",
        "America is the only country that went from barbarism to decadence without civilization in between. -Oscar Wilde",
        "Anyone who lives within their means suffers from a lack of imagination. -Oscar Wilde"
      ]
      quote = (rand(0)*quotes.size).floor
      quotes[quote]
    end
  end

  def respond io
    begin
      request = Request.new(io.gets)
      request.respond.each do |line|
        io << line
      end
    rescue Exception => e
      io.print "#{HTTP_VERSION} 500/Internal Server Error\r\n"
      io.print "Date: #{Time.now}\r\n"
      io.print "Content-Type: text/html\r\n"
      #io.print "ERROR: #{e.to_s}\r\n\r\n"
      io << Pathname.new(FILE_500).read
    end
  end
end

class BasicGServer < GServer
  include BasicServer
  def serve(io)
    respond(io)
  end
end

class BasicTCPServer
  include BasicServer
  def initialize(port, num_threads=10)
    @port = port
    @num_threads = num_threads
  end
  def start
    server = TCPServer.new(@port)
    threads = ThreadGroup.new
    while session = server.accept
      # Aaron said he thinks this is okay.
      if threads.list.size <= @num_threads
        t = Thread.new(session) do |my_session|
          respond(my_session)
          my_session.close
        end
        threads.add(t)
      end
    end
  end
end